cheri-c-programming.tex

```diff
@@ -1508,6 +1506,37 @@ \section{CHERI compiler warnings and errors}
   invoke_cb(&callback, (void*)(uintptr_t)callback_data); // no warning
 }
 \end{clisting}
+
+\subsection*{\texttt{alignment of '<type>' is less than the required capability alignment}}
+%\begin{compilerwarning}
+%alignment (<N>) of '<type>' is less than the required capability alignment
+%\end{compilerwarning}
+This warning is triggered when packed structures contain pointers.
+As mentioned in \cref{sec:restricted-capability-locations}, pointers must always be aligned to a the size of a CHERI capability (16 bytes for a 64-bit architecture).
+This warning can be triggered by code that attempts to align pointers to at least 8 bytes (e.g.\ for compatibility between 32- and 64-bit architectures). For example:
+\begin{clisting}[frame=single]
+struct AtLeast8ByteAlignedBad {
+  void *data;
+} __attribute__((packed, aligned(8)));
+// Compiler warns: alignment (8) of 'struct AtLeast8ByteAligned' is less than the required capability alignment (16})
```

Comment:
I know it's verbose, but I think I would actually have the source code snippets and then a preformatted box below that with the actual compiler warning (I think it's fine if it's "example.c" as the filename each time in the warning) showing the full warning output.

Reply:
I'm not particularly keen on having comments with the compiler warning in the source so I'll give that a try.